v1/tarball/image.go

```diff
@@ -15,17 +15,203 @@
+}
+
+func (i *image) MediaType() (types.MediaType, error) {
+	panic("not implemented")
+}
+
+// Manifest returns the v1.Manifest for the specified image.
+// This method memoizes the result, avoiding repeated reads from the tarball.
+func (i *image) Manifest() (*v1.Manifest, error) {
+	if i.manifest == nil {
+		if err := i.loadManifest(); err != nil {
+			return nil, err
+		}
+	}
+	return i.manifest, nil
+}
+
+// singleManifest is the struct used inside the `manifest.json` file of a `docker save` tarball.
+type singleManifest struct {
+	Config   string
+	RepoTags []string
+	Layers   []string
+}
+
+// tarManifest is the struct used to represent a single image inside a `docker save` tarball.
+type tarManifest []singleManifest
+
+func (i *image) loadManifest() error {
+	mfstBytes, err := extractFileFromTar(i.path, "manifest.json")
+	if err != nil {
+		return err
+	}
+
+	tarMfst := tarManifest{}
+	if err := json.Unmarshal(mfstBytes, &tarMfst); err != nil {
+		return err
+	}
+	if len(tarMfst) != 1 {
+		return errors.New("tarball must contain only a single image to be used with tarball.Image")
+	}
+
+	cfgBytes, err := extractFileFromTar(i.path, tarMfst[0].Config)
+	if err != nil {
+		return err
+	}
+
+	cfg, err := v1.ParseConfigFile(cfgBytes)
+	if err != nil {
+		return err
+	}
+	i.config = cfg
+
+	manifest := v1.Manifest{
+		SchemaVersion: 2,
+		MediaType:     types.DockerManifestSchema2,
+		Config: v1.Descriptor{
+			MediaType: types.DockerConfigJSON,
+			Size:      int64(len(cfgBytes)),
+			Digest:    v1.SHA256(string(cfgBytes)),
+		},
+	}
+
+	for _, l := range tarMfst[0].Layers {
+		// TODO(dlorenc): support compressed layers.
+		uncompressed, err := extractFileFromTar(i.path, l)
+		if err != nil {
+			return err
+		}
+		layer, err := compress(uncompressed)
+		if err != nil {
+			return err
+		}
+
+		manifest.Layers = append(manifest.Layers, v1.Descriptor{
+			MediaType: types.DockerLayer,
+			Size:      int64(len(layer)),
+			Digest:    v1.SHA256(string(layer)),
+		})
+	}
+	i.manifest = &manifest
+	return nil
+}
+
+func compress(l []byte) ([]byte, error) {
+	var buf bytes.Buffer
+	gz := gzip.NewWriter(&buf)
+	defer gz.Close()
+	if _, err := gz.Write(l); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+func extractFileFromTar(tarPath string, filePath string) ([]byte, error) {
+	f, err := os.Open(tarPath)
+	defer f.Close()
+	if err != nil {
+		return nil, err
+	}
+	tf := tar.NewReader(f)
+	for {
+		hdr, err := tf.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		if hdr.Name == filePath {
+			var buf bytes.Buffer
+			_, err := io.Copy(&buf, tf)
+			if err != nil {
+				return nil, err
+			}
+			return buf.Bytes(), nil
+		}
+	}
+	return nil, fmt.Errorf("file %s not found in tar", filePath)
+}
+
+func (i *image) ConfigFile() (*v1.ConfigFile, error) {
+	if i.config == nil {
+		if err := i.loadManifest(); err != nil {
+			return nil, err
+		}
+	}
+	return i.config, nil
```

Comment:
Accessing this technically shouldn't require loading the manifest IIRC.  Be lazier @dlorenc!!! ;-)

Reply:
Why not? You need to parse the manifest to figure out which blob is the config, right?

Reply:
I guess we can just guess that it's the file that ends with ".json" in the single manifest case?

Reply:
Ugh, sorry.  My turn to be confused by the naming of this file.